Rework pipeline data structure

qwerty287 opened the pull request:
supersedes https://github.com/woodpecker-ci/woodpecker/pull/3421
closes https://github.com/woodpecker-ci/woodpecker/issues/2971
closes https://github.com/woodpecker-ci/woodpecker/issues/4677
closes https://github.com/woodpecker-ci/woodpecker/issues/4195

Currently, our pipeline model has fields like the message, title etc., but every field is used differently, dependending on the event.

This PR changes this and splits these reused fields to have a more clear pipeline model.

The model now has fields like:
- commit (containing sha, author, forge url and message)
- pull request (index, title, from fork, labels)
- release title

and similar.

Advantages:
1. cleaner model and less differently used fields
2. fix the problem that there is a difference between commit and pipeline author (pipeline author = pusher) #2971 
3. store and show the commit message for every pipeline
4. store and show the context of a pipeline (PR title, deployment target, release title etc.) and show it

Especially 3 and 4 and are the enhancement parts here, as we can now store both context and commit message and not only one.

qwerty287 committed:
Do some work to refactor the pipeline model

qwerty287 committed:
continue - finish pipeline model

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
finish forge loaders

qwerty287 committed:
make author a pure commit author

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
migrate cli & web, add migration

qwerty287 committed:
rename to "Labels" only

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
fix feed

qwerty287 committed:
fix metadata panic

qwerty287 committed:
fix test panics

qwerty287 committed:
fix tests

qwerty287 committed:
fix gitea tag url

qwerty287 committed:
fix lint

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
Add migration docs and basic UI

pat-s commented:
@qwerty287 Why is that breaking? Because of the env var change?

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
resolve gitlab tag url

qwerty287 committed:
fix github tag url

qwerty287 committed:
sprintf

qwerty287 committed:
gitlab: load correct release author

qwerty287 committed:
fix web lint

qwerty287 committed:
try to fix mysql/postgres

xoxys commented:
Can you add a bit more context about what has changed and how it fixes the linked issue?

xoxys commented:
Thank you! ❤️

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
use tag messages

qwerty287 committed:
rename release title

xoxys commented:
Will finally take a look this evening.

xoxys commented:
After switching to this branch I get 
![image](https://github.com/user-attachments/assets/36f9558f-bfae-4fb4-9a82-0dc94d8a161d)

xoxys commented:
![image](https://github.com/user-attachments/assets/c08ed7ce-0cbc-47ad-85d1-0d30b8c9fb77)

I would stick to a two line layout, even if the second line would be empty. The pipeline id is not show for me.

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
fix feed

qwerty287 committed:
use flex

qwerty287 committed:
use trim

xoxys commented:
> I'm not sure if I'd do this.

True. But I would prefer to not change the layout depending on the event. Ill try to find another approach if not Im also fine with the current way.

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
Fix test

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
disable header linting

langecode commented on server/forge/bitbucketdatacenter/convert.go:
```diff
@@ -111,18 +112,19 @@ func convertRepositoryPushEvent(ev *bb.RepositoryPushEvent, baseURL string) *mod
+		Ref:         fmt.Sprintf("refs/pull-requests/%d/from", ev.PullRequest.ID),
+		PullRequest: convertPullRequest(&ev.PullRequest),
+		// TODO should link to the Pr
+		ForgeURL: fmt.Sprintf("%s/projects/%s/repos/%s/commits/%s", baseURL, ev.PullRequest.Source.Repository.Project.Key, ev.PullRequest.Source.Repository.Slug, ev.PullRequest.Source.Latest),
```
The link to the PR in the Bitbucket UI would be something like:

```
fmt.Sprintf("%s/projects/%s/repos/%s/pull-requests/%d", baseURL, ev.PullRequest.Source.Repository.Project.Key, ev.PullRequest.Source.Repository.Slug, ev.PullRequest.ID)
```

langecode commented on server/forge/bitbucketdatacenter/bitbucketdatacenter.go:
```diff
@@ -364,22 +364,28 @@ func (c *client) BranchHead(ctx context.Context, u *model.User, r *model.Repo, b
+	cm := commits[0]
+
+	return &model.Commit{
+		SHA:      cm.ID, // TODO check id or displayid?
```
`cm.ID` is the full sha - the `displayID` contains a shortened sha.

langecode commented on server/forge/bitbucketdatacenter/bitbucketdatacenter.go:
```diff
@@ -364,22 +364,28 @@ func (c *client) BranchHead(ctx context.Context, u *model.User, r *model.Repo, b
 		return nil, fmt.Errorf("unable to create bitbucket client: %w", err)
 	}
-	branches, _, err := bc.Projects.SearchBranches(ctx, r.Owner, r.Name, &bb.BranchSearchOptions{Filter: b})
+	commits, _, err := bc.Projects.SearchCommits(ctx, r.Owner, r.Name, &bb.CommitSearchOptions{
```
I'll see if I can do this API call to Bitbucket to verify that the `Until` property actually returns what we expect. The `b` variable will contain the branch name not the ref?

langecode reviewed (commented):
added a few comments to the bitbucket datacenter implementation.

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
pr url

qwerty287 committed:
solve comment

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
fix todo on bb url

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
escape ref

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

anbraten commented:
I like the change especially as it will help is in the future with adding PRs like #5174. 
Two suggestions:
- Extract the UI changes to a separate PR as they often involve some more discussions regarding the layout, etc
- Keep the current model structure for the api endpoints so this one could be merged soonishly without being breaking (therefore we could create a model that extends the new one and map the old properties accordingly)

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
fix  build

anbraten commented:
> I can remove the ui changes, but why can't we just do 4.0 and merge this as is with the breaking changes?

Breaking it without deprecating it in a previous release makes it extremely difficult for larger instances / teams to adopt to those api changes. Especially the pipeline apis are used by some external tools.

qwerty287 committed:
make non-breaking

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
remove ui changes

qwerty287 committed:
fix prettier

qwerty287 committed:
generate

qwerty287 committed:
fix tests

qwerty287 committed:
fix lint

qwerty287 committed:
openapi

6543 commented on pipeline/frontend/metadata/environment.go:
```diff
@@ -143,13 +142,16 @@ func (m *Metadata) Environ() map[string]string {
 	}
 
+	// TODO Deprecated, remove in next major
+	setNonEmptyEnvVar(params, "CI_COMMIT_AUTHOR_AVATAR", pipeline.Avatar)
```
I just wonder if CI_PIPELINE_AVATAR is interesting for people to be used?

6543 commented on server/api/pipeline.go:
```diff
@@ -79,25 +79,21 @@ func CreatePipeline(c *gin.Context) {
-		Avatar:  user.Avatar,
-		Message: "MANUAL PIPELINE @ " + opts.Branch,
+		Event:  event,
+		Commit: commit,
```
timestamp is now missing ?!?

6543 replied:
timestamp is now proper in commit ...

6543 commented on server/model/pipeline.go:
```diff
@@ -20,46 +20,85 @@ import (
-	Refspec             string                 `json:"refspec"                 xorm:"refspec"`
-	Title               string                 `json:"title"                   xorm:"title"`
-	Message             string                 `json:"message"                 xorm:"TEXT 'message'"`
-	Timestamp           int64                  `json:"timestamp"               xorm:"'timestamp'"`
```
timestamp got removed why?

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

anbraten commented on docs/src/pages/migrations.md:
```diff
@@ -1,10 +1,19 @@
+
+### API changes
+
+- Changed the pipeline model to have different objects for different event metadata (e.g. pull request title)
```
Might be best to list the changes. Helped me with understanding them while reviewing as  well.

```suggestion
- Changed the pipeline model to use nested objects grouped based on the event (e.g. instead of a generic `title` it now uses `pr.title`)

`author` => `commit.author`
`deploy_to` => `deployment.target`
`deploy_task` => `deployment.task`
`commit` (sha) => `commit.sha`
`title` => `release.title` (for release events) & pr.title (for pull-request events)
`message` => `commit.message`
`timestamp` => `created` (somehow dropped the value, was this the commit timestamp before?)
`sender` => `author`
`avatar` => `author_avatar`
`author_email` => `commit.author.email`
`pr_labels` => `pr.labels`
`is_prerelease` => `is_prerelease` (not `release.is_prerelease` ?)
??? => `release_tag_title` (not `release.tag_title`? was this the normal `title` before?)
`from_fork` => `pr.from_fork`
```

anbraten commented on docs/src/pages/migrations.md:
```diff
@@ -83,6 +92,38 @@ The following syntax deprecations will now result in an error:
 - `platform:` ([#3916](https://github.com/woodpecker-ci/woodpecker/pull/3916))
 - `branches:` ([#3916](https://github.com/woodpecker-ci/woodpecker/pull/3916))
 
+#### Workflow syntax changes
```
Where is this coming from? Should we add it in a separate PR?

anbraten commented on pipeline/frontend/metadata/types.go:
```diff
@@ -52,9 +52,13 @@ type (
 		DeployTo   string `json:"target,omitempty"`
 		DeployTask string `json:"task,omitempty"`
-		Commit     Commit `json:"commit,omitempty"`
+		Commit     Commit `json:"commit"`
```
Why not `omitempty`?

anbraten commented on server/api/pipeline.go:
```diff
@@ -638,7 +641,7 @@ func PostPipeline(c *gin.Context) {
 		}
 
-		pl.DeployTo = c.DefaultQuery("deploy_to", pl.DeployTo)
+		pl.Deployment.Target = c.DefaultQuery("deploy_to", pl.Deployment.Target)
```
This property is called `deployment.target` now while the query parameter is `deploy_to`. Should use one term to be consistent with a deprecation and fallback if necessary.

anbraten commented on server/model/pipeline.go:
```diff
@@ -20,46 +20,85 @@ import (
+	ap := &APIPipeline{
+		Pipeline:  p,
+		Commit:    p.Commit.SHA,
+		Title:     p.Commit.Message,
```
This should be using sth like the `pr.title` or `release_tag_title` or `cron` based on the event to be backwards compatible.

anbraten commented on web/src/components/repo/pipeline/PipelineItem.vue:
```diff
@@ -54,7 +54,7 @@
         <div class="flex min-w-0 items-center space-x-2">
           <Icon name="commit" />
-          <span class="truncate">{{ pipeline.commit.slice(0, 10) }}</span>
+          <span class="truncate">{{ pipeline.commit_pipeline.sha.slice(0, 10) }}</span>
```
As the api is now backwards compatible, I would suggest to extract these changes to a followup PR, so we can check such fundamental changes one-by-one.

anbraten replied:
Would favor to do this in small steps as it would allow us to deploy this first and check if the api is still working as before and later on we could check the ui update using the new data fields.

anbraten reviewed (commented):
Nice work. Code mainly lgtm (haven't check on the mappings from the forges to the model too much yet). As those changes are quite fundamental I would suggest to extract all possible parts like the UI and env var changes from this first PR, so we can easily test and find the change if an error occurs.

qwerty287 committed:
Update server/store/datastore/migration/026_update_pipeline_messages.go

Co-authored-by: Anbraten <6918444+anbraten@users.noreply.github.com>

qwerty287 committed:
fix merge issue

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
rename deploy param

qwerty287 committed:
add cli flags back

qwerty287 committed:
change message based on event

qwerty287 committed:
fix build issues

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
revert web changes

qwerty287 committed:
Add release field

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
openapi

anbraten commented on server/forge/forgejo/fixtures/HookPush.json:
```diff
@@ -18,6 +18,19 @@
       "modified": ["app/controller/application.rb"]
     }
   ],
+  "head_commit": {
```
Has this been added to forgejo?

anbraten commented on server/forge/forgejo/forgejo.go:
```diff
@@ -643,34 +667,75 @@ func (c *Forgejo) getChangedFilesForPR(ctx context.Context, repo *model.Repo, in
 }
 
-func (c *Forgejo) getTagCommitSHA(ctx context.Context, repo *model.Repo, tagName string) (string, error) {
+func (c *Forgejo) getTagMessage(ctx context.Context, repo *model.Repo, tagName string) (*model.Commit, string, error) {
```
```suggestion
func (c *Forgejo) getTagCommitAndMessage(ctx context.Context, repo *model.Repo, tagName string) (*model.Commit, string, error) {
```

anbraten commented on server/forge/forgejo/forgejo.go:
```diff
@@ -489,13 +494,32 @@ func (c *Forgejo) Hook(ctx context.Context, r *http.Request) (*model.Repo, *mode
+			pipeline.Commit = sha
+		case pipeline.Event == model.EventTag:
+			tagName := strings.Split(pipeline.Ref, "/")[2]
+			sha, msg, err := c.getTagMessage(ctx, repo, tagName)
```
```suggestion
			commit, tagMsg, err := c.getTagMessage(ctx, repo, tagName)
```

anbraten commented on server/forge/forgejo/forgejo.go:
```diff
@@ -489,13 +494,32 @@ func (c *Forgejo) Hook(ctx context.Context, r *http.Request) (*model.Repo, *mode
+		switch {
+		case pipeline.Event == model.EventRelease && pipeline.Commit.SHA == "":
+			tagName := strings.Split(pipeline.Ref, "/")[2]
+			sha, _, err := c.getTagMessage(ctx, repo, tagName)
```
```suggestion
			commit, _, err := c.getTagMessage(ctx, repo, tagName)
```

anbraten commented on server/forge/forgejo/forgejo.go:
```diff
@@ -489,13 +494,32 @@ func (c *Forgejo) Hook(ctx context.Context, r *http.Request) (*model.Repo, *mode
+				return nil, nil, err
+			}
+			pipeline.Commit = sha
+			if pipeline.Commit.Message != msg {
```
Why should this only be set if its not the same?

anbraten commented on server/forge/forgejo/forgejo.go:
```diff
@@ -643,34 +667,75 @@ func (c *Forgejo) getChangedFilesForPR(ctx context.Context, repo *model.Repo, in
 	}
 
-	return tag.Commit.SHA, nil
+	commit, err := c.getCommitFromSHA(ctx, user, repo, tag.Commit.SHA)
```
Might be cleaner to not include this call directly in here, but call it in the hook directly

anbraten commented on server/forge/forgejo/forgejo.go:
```diff
@@ -643,34 +667,75 @@ func (c *Forgejo) getChangedFilesForPR(ctx context.Context, repo *model.Repo, in
+	return commit, tag.Message, err
+}
+
+func (c *Forgejo) getCommitFromSHAStore(ctx context.Context, repo *model.Repo, sha string) (*model.Commit, error) {
```
This name is kinda confusing as it does not get the sha from the store, but loads some other data from it instead to call the forge api underneath

anbraten replied:
No nice idea, would change the helpers scope in a followup and rename in that moment than.

anbraten commented on server/forge/gitea/gitea.go:
```diff
@@ -496,13 +497,32 @@ func (c *Gitea) Hook(ctx context.Context, r *http.Request) (*model.Repo, *model.
+		switch {
+		case pipeline.Event == model.EventRelease && pipeline.Commit.SHA == "":
+			tagName := strings.Split(pipeline.Ref, "/")[2]
+			sha, _, err := c.getTagMessages(ctx, repo, tagName)
```
same as for forgejo

anbraten commented on server/forge/gitea/helper.go:
```diff
@@ -267,6 +259,15 @@ func matchingHooks(hooks []*gitea.Hook, rawURL string) *gitea.Hook {
 	return nil
 }
 
+func convertPullRequests(from *gitea.PullRequest) *model.PullRequest {
```
```suggestion
func convertPullRequest(from *gitea.PullRequest) *model.PullRequest {
```

anbraten commented on server/forge/gitea/parse_test.go:
```diff
@@ -64,15 +64,21 @@ func TestGiteaParser(t *testing.T) {
-				Sender:       "6543",
-				Avatar:       "https://codeberg.org/avatars/09a234c768cb9bca78f6b2f82d6af173",
-				Email:        "6543@obermui.de",
+				Event: "push",
```
```suggestion
				Event: model.EventPush,
```

anbraten commented on server/forge/github/convert.go:
```diff
@@ -168,3 +178,15 @@ func convertLabels(from []*github.Label) []string {
+		ForgeURL: from.GetHTMLURL(),
+		Message:  from.GetMessage(),
+		Author: model.CommitAuthor{
+			Author: from.GetAuthor().GetName(),
```
This is actually a quite nice idea to use `author.name` instead of `author.author`

anbraten commented on server/forge/github/github.go:
```diff
@@ -610,13 +604,21 @@ func (c *client) Hook(ctx context.Context, r *http.Request) (*model.Repo, *model
+			}
+			pipeline.Commit = sha
+		} else if pipeline.Event == model.EventDeploy || pipeline.Event == model.EventPull || pipeline.Event == model.EventPullClosed {
+			sha, err := c.getCommitFromSHA(ctx, repo, pipeline.Commit.SHA)
```
```suggestion
			commit, err := c.getCommitFromSHA(ctx, repo, pipeline.Commit.SHA)
```

anbraten commented on server/forge/github/github.go:
```diff
@@ -610,13 +604,21 @@ func (c *client) Hook(ctx context.Context, r *http.Request) (*model.Repo, *model
+	if pipeline != nil {
+		if pipeline.Event == model.EventRelease && pipeline.Commit.SHA == "" {
+			tagName := strings.Split(pipeline.Ref, "/")[2]
+			sha, err := c.getTagCommitSHA(ctx, repo, tagName)
```
```suggestion
			commit, err := c.getTagCommitSHA(ctx, repo, tagName)
```

anbraten commented on server/forge/github/github.go:
```diff
@@ -667,21 +669,20 @@ func (c *client) loadChangedFilesFromPullRequest(ctx context.Context, pull *gith
 }
 
-func (c *client) getTagCommitSHA(ctx context.Context, repo *model.Repo, tagName string) (string, error) {
+func (c *client) getTagCommitSHA(ctx context.Context, repo *model.Repo, tagName string) (*model.Commit, error) {
```
Similar to gitea / forgejo I would improve those two helpers to do less "magic" and rather one thing like one returning the sha and one the commit from a sha or similar.

```suggestion
func (c *client) getCommitAndMessageFromTag(ctx context.Context, repo *model.Repo, tagName string) (*model.Commit, error) {
```

anbraten commented on server/forge/github/parse.go:
```diff
@@ -85,32 +85,28 @@ func parsePushHook(hook *github.PushEvent) (*model.Repo, *model.Pipeline) {
 		// event we'll never know!
 		pipeline.Event = model.EventTag
 		pipeline.ChangedFiles = nil
-		// For tags, if the base_ref (tag's base branch) is set, we're using it
```
Shouldn't we keep this?

anbraten commented on server/forge/github/github.go:
```diff
@@ -610,13 +604,21 @@ func (c *client) Hook(ctx context.Context, r *http.Request) (*model.Repo, *model
-		if err != nil {
-			return nil, nil, err
+	if pipeline != nil {
+		if pipeline.Event == model.EventRelease && pipeline.Commit.SHA == "" {
```
Maybe a check if message is empty but sha is set to load the commit details would be cleaner and could help loading the corresponding data always if possible. Cculd be especially helpful when adding new events later on.

anbraten commented on server/model/feed.go:
```diff
@@ -17,21 +17,42 @@ package model
+	PullRequest *PullRequest `json:"pull_request,omitempty"      xorm:"json 'pipeline_pr'"`
+	Release     *Release     `json:"release,omitempty" xorm:"json 'pipeline_release'"`
+	// TODO change json to 'commit' in next major
+	Commit *Commit `json:"commit_pipeline,omitempty"   xorm:"json 'pipeline_commit'"`
```
Am I right that we will drop `pipeline_commit` in favor of `commit` in the next major or would we rename again?

anbraten commented on server/model/pipeline.go:
```diff
@@ -20,46 +20,97 @@ import (
+	Ref          string       `json:"ref"                         xorm:"ref"`
+	Refspec      string       `json:"refspec"                     xorm:"refspec"`
+	ForgeURL     string       `json:"forge_url"                   xorm:"forge_url"`
+	Author       string       `json:"author"                      xorm:"author"`
```
```suggestion
	Author       string       `json:"author"                      xorm:"author"` // The user sending the webhook data or triggering the pipeline event
```

anbraten commented on server/model/pipeline.go:
```diff
@@ -78,3 +129,14 @@ type PipelineOptions struct {
+
+type Release struct {
+	IsPrerelease bool   `json:"is_prerelease,omitempty"`
+	TagTitle     string `json:"release_tag_title,omitempty"`
```
```suggestion
	TagTitle     string `json:"tag_title,omitempty"`
```

Do we want to know the tag title or the release title here?

anbraten commented on woodpecker-go/woodpecker/types.go:
```diff
@@ -96,32 +96,54 @@ type (
+		Started     int64            `json:"started_at"`
+		Finished    int64            `json:"finished_at"`
+		Deployment  *Deployment      `json:"deployment"`
+		Commit      *Commit          `json:"commit"`
```
`pipeline_commit`?

qwerty287 committed:
add deprecated api param to docs

qwerty287 committed:
renamings

qwerty287 committed:
gitea as well

qwerty287 committed:
some more fixes

qwerty287 committed:
Update docs/src/pages/migrations.md

Co-authored-by: Anbraten <6918444+anbraten@users.noreply.github.com>

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
add separate tag and release title

qwerty287 committed:
Update server/model/pipeline.go

Co-authored-by: Anbraten <6918444+anbraten@users.noreply.github.com>

qwerty287 committed:
fix sdk

qwerty287 committed:
remove env note

qwerty287 committed:
rename func

anbraten commented on woodpecker-go/woodpecker/types.go:
```diff
@@ -96,32 +96,54 @@ type (
+	}
+
+	CommitAuthor struct {
+		Author string `json:"author"`
```
```suggestion
		Name string `json:"name"`
```

anbraten commented on docs/docs/20-usage/50-environment.md:
```diff
@@ -77,7 +77,6 @@ This is the reference list of all environment variables available to your pipeli
 | `CI_COMMIT_MESSAGE`              | commit message                                                                                                     | `Initial commit`                                                                           |
 | `CI_COMMIT_AUTHOR`               | commit author username                                                                                             | `john-doe`                                                                                 |
 | `CI_COMMIT_AUTHOR_EMAIL`         | commit author email address                                                                                        | `john-doe@example.com`                                                                     |
-| `CI_COMMIT_AUTHOR_AVATAR`        | commit author avatar                                                                                               | `https://git.example.com/avatars/5dcbcadbce6f87f8abef`                                     |
```
Should be in the migration notes or reverted and be changed as part  of #5227?!

anbraten commented on server/store/datastore/migration/026_update_pipeline_messages.go:
```diff
@@ -0,0 +1,159 @@
+			Description string `json:"description"`
+		}
+
+		type pipelines struct {
```
```suggestion
		type pipeline struct {
```

anbraten commented on server/store/datastore/migration/026_update_pipeline_messages.go:
```diff
@@ -0,0 +1,159 @@
+			Deployment      *deployment  `xorm:"json 'deployment'"`
+			PullRequest     *pullRequest `xorm:"json 'pr'"`
+			Cron            string       `xorm:"cron"`
+			ReleaseTagTitle string       `xorm:"release_tag_title"`
```
would it help to define separate types for old and new pipelines?

anbraten commented on server/store/datastore/migration/026_update_pipeline_messages.go:
```diff
@@ -0,0 +1,159 @@
+
+var updatePipelineMessages = xormigrate.Migration{
+	ID: "update-pipeline-messages",
+	MigrateSession: func(sess *xorm.Session) error {
```
Before merging we should do one final test using a copy of our production db.

anbraten reviewed (commented):
Hope I detected all remaining pieces. Should be almost rdy to go.

qwerty287 committed:
rename

qwerty287 committed:
fix migration types

qwerty287 committed:
remove env var docs change

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
revert sdk

qwerty287 committed:
some tag improvements

qwerty287 committed:
add own tag title

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
openapi

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
generate openapi

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
Add deprecation note to old sdk fields

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
openapi

anbraten committed:
adjust pipeline properties

6543 committed:
Merge branch 'main' into pl-commit-pr

6543 commented on server/api/pipeline_test.go:
```diff
@@ -37,6 +37,7 @@ var fakePipeline = &model.Pipeline{
 	ID:     2,
 	Number: 2,
 	Status: model.StatusSuccess,
+	Commit: &model.Commit{},
```
why?

qwerty287 replied:
Tests fail otherwise with nil pointer. This way e.g. `Commit.SHA` is defined as empty string.

6543 commented on server/model/pipeline.go:
```diff
@@ -32,34 +32,92 @@ type Pipeline struct {
-	IsPrerelease        bool                   `json:"is_prerelease,omitempty" xorm:"is_prerelease"`
-	FromFork            bool                   `json:"from_fork,omitempty"     xorm:"from_fork"`
+	Deployment          *Deployment            `json:"deployment,omitempty"    xorm:"json 'deployment'"`
+	PullRequest         *PullRequest           `json:"pull_request,omitempty"  xorm:"json 'pull_request'"`
```
yes as data-blob you can store it ... but you can then no longer do standart sql querys e.g. for filtering purpose.

we should filter more stuff in the backend and not the frontend as we do now ...

6543 commented on server/forge/bitbucket/internal/client.go:
```diff
@@ -206,6 +207,13 @@ func (c *Client) GetBranchHead(owner, name, branch string) (*Commit, error) {
 	return out.Values[0], nil
 }
 
+func (c *Client) GetCommit(owner, name, sha string) (*Commit, error) {
```
I'm adding this addition into https://github.com/woodpecker-ci/woodpecker/pull/5410

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

6543 committed:
Merge branch 'main' into pl-commit-pr

6543 committed:
Merge branch 'main' into pl-commit-pr

6543 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
Merge commit 'e9c545e25b997bd9e8f99d5c18f4ecc1b3f3c301' into pl-commit-pr

qwerty287 committed:
generate

qwerty287 committed:
Merge commit '388557d94aa9fe06841a207eba4ffa4ba0ae5a05' into pl-commit-pr

qwerty287 committed:
fmt

qwerty287 committed:
fix tests

qwerty287 committed:
Merge commit 'c8b0be29fe11306fca25421aabb89501c3c95990' into pl-commit-pr

qwerty287 committed:
Merge commit '9fa82d3a1c0093a3788eb01ead7d75683519dcfd' into pl-commit-pr

qwerty287 committed:
Merge commit '22ffe9373468961b176fe52829a612b44408395a' into pl-commit-pr

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
openapi

qwerty287 committed:
fix lint

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
fix merge

qwerty287 commented:
@woodpecker-ci/maintainers could I ask again for reviews here? This is open for almost a year now. And it simplifies the pipeline structure a lot. (Also feel free to directly modify the branch if there's something to fix)

anbraten commented:
Thanks for updating to main again. I will try to have a look at it and will do some testing. Could you try to test with a copy of our ci.woodpecker-ci.org db as well. Last time it had some hickups with the migrations.

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
openapi

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

6543 committed:
Merge branch 'main' into pl-commit-pr

6543 committed:
Merge branch 'main' into pl-commit-pr

6543 committed:
Merge branch 'main' into pl-commit-pr

6543 committed:
Merge branch 'main' into pl-commit-pr

6543 commented on server/api/pipeline.go:
```diff
@@ -78,26 +78,27 @@ func CreatePipeline(c *gin.Context) {
+		c.JSON(http.StatusOK, pl.ToAPIModel())
 	} else {
-		c.JSON(http.StatusOK, pl)
+		c.String(http.StatusOK, "nothing to run")
```
api expects json ... we should not change that

esp. in a non error paths

6543 committed:
Apply suggestion from @6543

qwerty287 commented:
I'll try to check again in the next days.

6543 committed:
Merge branch 'main' into pl-commit-pr

6543 commented:
efen with xorm you can write raw sql querys, and postgress allows json filters... so it's not a huge tradeoff, we could state, that for big installation we only recomed tjat database and for the rest we filter via go-code

we just should make that deccission in consensus

https://www.postgresql.org/docs/current/functions-json.html

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
fix err name

qwerty287 committed:
gen and format

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
openapi

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
fix merge

qwerty287 committed:
openapi

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

6543 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

6543 commented on server/pipeline/create.go:
```diff
@@ -42,13 +42,9 @@ func Create(ctx context.Context, _store store.Store, repo *model.Repo, pipeline
 
 	if pipeline.Event == model.EventPush || pipeline.IsPullRequest() {
-		skipMatch := skipPipelineRegex.FindString(pipeline.Message)
+		skipMatch := skipPipelineRegex.FindString(pipeline.Commit.Message)
```
-> https://github.com/woodpecker-ci/woodpecker/pull/6437

just did the refactoring

6543 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

6543 committed:
Merge branch 'main' into pl-commit-pr

6543 commented:
well about the DB normalisation ... i think we can still do that later ... so to move this forward let that as is and if needed we can always have a followup.

I would suggest to merge this avter the v3.14 release so we have time to pick up issues

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

6543 committed:
Merge branch 'main' into pl-commit-pr

6543 committed:
adapt

6543 committed:
Merge branch 'main' into pl-commit-pr

6543 commented on server/api/pipeline.go:
```diff
@@ -90,19 +90,15 @@ func CreatePipeline(c *gin.Context) {
-		Event:     event,
-		Commit:    commit.SHA,
-		Branch:    opts.Branch,
-		Timestamp: time.Now().UTC().Unix(),
```
the timestamp where did that info go?

6543 replied:
cah created...

6543 commented on server/model/pipeline.go:
```diff
@@ -73,13 +80,46 @@ func (Pipeline) TableName() string {
+		}
+		ap.Message = "created release " + p.TagTitle
+	case EventManual:
+		ap.Message = "MANUAL PIPELINE" + p.Branch
```
```suggestion
		ap.Message = "MANUAL PIPELINE @ " + p.Branch
```

6543 committed:
Apply suggestion from @6543

6543 commented:
the only thing blocking is that i have to review the migration part ... but it looks costly ...

qwerty287 commented:
We still need to discuss how to work with the migration time. As you said it will take days on instances like codeberg… https://github.com/woodpecker-ci/woodpecker/pull/4626#issuecomment-3808027801

Also I'd like to finish the discussion on the database model before merging, otherwise we basically revert that at DB level later… https://github.com/woodpecker-ci/woodpecker/pull/4626#issuecomment-4204265522

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
fix merge

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
fix year

6543 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

6543 commented on server/forge/gitlab/convert.go:
```diff
@@ -275,22 +276,30 @@ func convertPushHook(hook *gitlab.PushEvent) (*model.Repo, *model.Pipeline, erro
 	files := make([]string, 0, len(hook.Commits)*4)
 	for _, cm := range hook.Commits {
 		if hook.After == cm.ID {
+<<<<<<< HEAD
```
merge conflict relicts got in with the last merge

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
fix merge and test

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

6543 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
Merge branch 'main' into pl-commit-pr

qwerty287 committed:
fix test

qwerty287 committed:
openapi

6543 committed:
Merge branch 'main' into pl-commit-pr

claude committed:
fix(e2e): adapt concurrency test to reworked Pipeline data structure

Pipeline.Commit is now *model.Commit (not string) and Sender was removed
in favor of Author. Update the struct literal accordingly.

6543 committed:
Merge branch 'main' into pl-commit-pr

6543 committed:
Merge branch 'main' into pl-commit-pr

6543 commented on server/model/pipeline.go:
```diff
@@ -21,50 +21,57 @@ import (
+	Updated             int64                   `json:"updated"                 xorm:"'updated' NOT NULL DEFAULT 0 updated"`
+	Started             int64                   `json:"started"                 xorm:"started"`
+	Finished            int64                   `json:"finished"                xorm:"finished"`
+	Commit              *Commit                 `json:"commit_pipeline"         xorm:"json 'commit'"` // TODO change json to 'commit' in next major
```
this colides

6543 committed:
migration: less copy

6543 committed:
start migration fix

6543 committed:
migrate milestones tooo

6543 committed:
api convert no nil issue

6543 committed:
be save with nil

6543 committed:
bulk update

6543 committed:
Merge branch 'main' into pl-commit-pr

claude committed:
fix(server): guard nil Commit in skip-ci check

createPipeline dereferenced pipeline.Commit unconditionally for the
skip-ci message and SHA lookup. With the new pointer-based Commit field a
pipeline carrying no commit metadata would panic. Read the values behind
a nil check, consistent with the nil handling added in gated.go and
metadata.go.

6543 committed:
trim pull index same way as webui

6543 committed:
trim pull index same way as webui (2)

6543 commented on server/pipeline/create.go:
```diff
@@ -40,12 +40,18 @@ func Create(ctx context.Context, _store store.Store, repo *model.Repo, pipeline
+	}
+
+	if constraint.IsSkipCommitMessage(metadata.Event(pipeline.Event), commitMessage) {
+		ref := commitSHA
```
```suggestion
```

6543 commented on server/pipeline/create.go:
```diff
@@ -40,12 +40,18 @@ func Create(ctx context.Context, _store store.Store, repo *model.Repo, pipeline
 			ref = pipeline.Ref
 		}
-		log.Debug().Str("repo", repo.FullName).Msgf("ignoring pipeline as skip-ci was found in the commit (%s) message '%s'", ref, pipeline.Message)
+		log.Debug().Str("repo", repo.FullName).Msgf("ignoring pipeline as skip-ci was found in the commit (%s) message '%s'", ref, commitMessage)
```
```suggestion
		log.Debug().Str("repo", repo.FullName).Msgf("ignoring pipeline as skip-ci was found in the commit (%s) message '%s'", commitSHA, commitMessage)
```

6543 committed:
Apply suggestions from code review

Co-authored-by: 6543 <6543@obermui.de>

6543 committed:
Merge branch 'main' into pl-commit-pr

claude committed:
fix(server): remove dead undefined ref in skip-ci branch

The pipeline data structure rework left a stale block referencing an
undeclared 'ref' variable inside the skip-commit-message branch. The
block assigned to 'ref' and then returned ErrFiltered immediately, so
the value was never used. It only survived as a leftover and broke the
build with 'undefined: ref'.

Removing it fixes the compile error that cascaded into the
cross-compile-server-preview, lint (typecheck), test and test-e2e
steps.

6543 committed:
fix merge conflict relicts

6543 committed:
prepare for merge

6543 committed:
Merge branch 'main' into pl-commit-pr

6543 committed:
fix merge conflict relicts

claude committed:
fix(server): drop stale avatar column from feed query

The pipeline data rework removed the pipelines.avatar column (replaced
by author_avatar), but feedTemplate still selected it, so
GetPipelineQueue failed with "no such column: pipelines.avatar" and
returned nothing. Also drop a duplicate pipeline_release select.

claude committed:
fix(migration): avoid nil deref migrating pull request draft

The 029 migration builds the new PullRequest field while reading
p.PullRequest.Draft, but that pointer is still nil at this point, so
TestMigrate panics. Read the draft flag from the old flat pr_draft
column instead.

6543 committed:
Merge branch 'main' into pl-commit-pr

6543 committed:
fix tests
